File: java-source/src/main/java/net/corda/training/flow/IOUTransferFlow.java
package net.corda.training.flow;

import co.paralleluniverse.fibers.Suspendable;
import net.corda.core.contracts.Command;
import net.corda.core.contracts.ContractState;
import net.corda.core.contracts.StateAndRef;
import net.corda.core.crypto.SecureHash;
import net.corda.core.flows.*;
import net.corda.core.node.services.vault.Builder;
import net.corda.core.node.services.vault.CriteriaExpression;
import net.corda.core.utilities.ProgressTracker;
import net.corda.core.contracts.UniqueIdentifier;
import net.corda.core.identity.AbstractParty;
import net.corda.core.identity.Party;
import net.corda.core.node.services.Vault;
import net.corda.core.node.services.vault.QueryCriteria;
import net.corda.core.transactions.SignedTransaction;
import net.corda.core.transactions.TransactionBuilder;
import net.corda.training.contract.IOUContract;
import net.corda.training.state.IOUState;

import javax.annotation.Signed;
import javax.validation.constraints.NotNull;
import java.security.PublicKey;
import java.util.*;
import java.util.stream.Collectors;

import static net.corda.core.contracts.ContractsDSL.requireThat;

public class IOUTransferFlow{

    @InitiatingFlow
    @StartableByRPC
    public static class InitiatorFlow extends FlowLogic<SignedTransaction> {

        private final UniqueIdentifier stateLinearId;
        private final Party newLender;

        public InitiatorFlow(UniqueIdentifier stateLinearId, Party newLender) {
            this.stateLinearId = stateLinearId;
            this.newLender = newLender;
        }

        @Suspendable
        public SignedTransaction call() throws FlowException {

            // Task 1

            QueryCriteria queryCriteria = new QueryCriteria.LinearStateQueryCriteria(null, new ArrayList<>(Arrays.asList(stateLinearId.getId())));
            Vault.Page<IOUState> query = getServiceHub().getVaultService().queryBy(IOUState.class, queryCriteria);
            IOUState iouState = query.getStates().get(0).getState().getData();

            IOUState iouStateCopy = iouState.withNewLender(newLender);

            List<PublicKey> requiredKeys = new ArrayList<>();
            requiredKeys.addAll(iouState.getParticipants().stream().map(AbstractParty::getOwningKey).collect(Collectors.toList()));
            requiredKeys.add(newLender.getOwningKey());

            final Command<IOUContract.Commands.Transfer> transferCommand = new Command<>(new IOUContract.Commands.Transfer(), requiredKeys);

            final Party notary = getServiceHub().getNetworkMapCache().getNotaryIdentities().get(0);
            final TransactionBuilder transactionBuilder = new TransactionBuilder(notary);
            transactionBuilder.addCommand(transferCommand);
            StateAndRef stateAndRef = query.getStates().get(0);
            transactionBuilder.addInputState(stateAndRef);
            transactionBuilder.addOutputState(iouStateCopy);

            transactionBuilder.verify(getServiceHub());
            SignedTransaction partiallySignedTransaction = getServiceHub().signInitialTransaction(transactionBuilder);

            // Task 2
            if (!getOurIdentity().equals(iouState.getLender())) throw new IllegalArgumentException("This flow must be run by the current lender.");

            // Task 3

            // Task 4
            Set<Party> participants = new HashSet<>();
            participants.add(iouState.getBorrower());
            participants.add(newLender);
            Set<FlowSession> flowSessions = participants.stream().map(this::initiateFlow).collect(Collectors.toSet());
            SignedTransaction signedTransaction = subFlow(new CollectSignaturesFlow(partiallySignedTransaction, flowSessions));

            // Task 5
            return subFlow(new FinalityFlow(signedTransaction, flowSessions));
        }
    }


    /**
     * This is the flow which signs IOU settlements.
     * The signing is handled by the [SignTransactionFlow].
     * Uncomment the initiatedBy annotation to facilitate the responder flow.
     */
    @InitiatedBy(IOUTransferFlow.InitiatorFlow.class)
    public static class Responder extends FlowLogic<SignedTransaction> {

        private final FlowSession otherPartyFlow;
        private SecureHash txWeJustSigned;
        public Responder(FlowSession otherPartyFlow) {
            this.otherPartyFlow = otherPartyFlow;
        }

        @Suspendable
        @Override
        public SignedTransaction call() throws FlowException {
            class SignTxFlow extends SignTransactionFlow {
                private SignTxFlow(FlowSession otherPartyFlow, ProgressTracker progressTracker) {
                    super(otherPartyFlow, progressTracker);
                }
                @Override
                @NotNull
                protected void checkTransaction(SignedTransaction stx) {
                    requireThat(require -> {
                        ContractState output = stx.getTx().getOutputs().get(0).getData();
                        require.using("This must be an IOU transaction", output instanceof IOUState);
                        return null;
                    });
                    txWeJustSigned = stx.getId();
                }
            }

            SignTxFlow signTxFlow = new SignTxFlow(otherPartyFlow, SignTxFlow.tracker());

            subFlow(signTxFlow);

            return subFlow(new ReceiveFinalityFlow(otherPartyFlow, txWeJustSigned));
        }

    }

}

File: kotlin-source/src/main/kotlin/net/corda/training/flow/IOUTransferFlow.kt
package net.corda.training.flow

import co.paralleluniverse.fibers.Suspendable
import net.corda.core.contracts.*
import net.corda.core.flows.*
import net.corda.core.identity.Party
import net.corda.core.node.services.queryBy
import net.corda.core.node.services.vault.QueryCriteria
import net.corda.core.transactions.SignedTransaction
import net.corda.core.transactions.TransactionBuilder
import net.corda.core.utilities.ProgressTracker
import net.corda.training.contract.IOUContract
import net.corda.training.state.IOUState

/**
 * This is the flow which handles transfers of existing IOUs on the ledger.
 * Gathering the counterparty's signature is handled by the [CollectSignaturesFlow].
 * Notarisation (if required) and commitment to the ledger is handled by the [FinalityFlow].
 * The flow returns the [SignedTransaction] that was committed to the ledger.
 */
@InitiatingFlow
@StartableByRPC
class IOUTransferFlow(val linearId: UniqueIdentifier, val newLender: Party) : FlowLogic<SignedTransaction>() {

    override val progressTracker = tracker()

    companion object {
        object CREATING : ProgressTracker.Step("Creating a transfer IOU transaction")
        object VERIFYING : ProgressTracker.Step("Verifying the IOU transfer")
        object SIGNING : ProgressTracker.Step("Signing the IOU transfer")
        object STORING : ProgressTracker.Step("Processing storage of IOU transfer") {
            override fun childProgressTracker() = FinalityFlow.tracker()
        }

        fun tracker() = ProgressTracker(CREATING, SIGNING, VERIFYING, STORING)
    }

    @Suspendable
    override fun call(): SignedTransaction {
        progressTracker.currentStep = CREATING
        val notary = serviceHub.networkMapCache.notaryIdentities.single()

        val inputStateAndRef = getInputStateAndRef(linearId)
        val (utx, participants) = createUnsignedTransaction(notary, inputStateAndRef)

        progressTracker.currentStep = VERIFYING
        utx.verify(serviceHub)

        progressTracker.currentStep = SIGNING
        val stx = serviceHub.signInitialTransaction(utx)

        val counterpartySessions = participants.filterNot { it == ourIdentity }.map { initiateFlow(it) }
        val ctx = subFlow(CollectSignaturesFlow(stx, counterpartySessions))

        return subFlow(FinalityFlow(ctx, counterpartySessions))
    }

    private fun getInputStateAndRef(linearId: UniqueIdentifier): StateAndRef<IOUState> {
        val linearStateCriteria =
            QueryCriteria.LinearStateQueryCriteria(linearId = listOf(linearId))
        return serviceHub.vaultService.queryBy<IOUState>(linearStateCriteria).states.single()
    }

    private fun createUnsignedTransaction(
        notary: Party,
        inputStateAndRef: StateAndRef<IOUState>
    ): Pair<TransactionBuilder, List<Party>> {
        val inputIOU = inputStateAndRef.state.data
        val participants = inputStateAndRef.state.data.participants + newLender

        if (inputIOU.lender != ourIdentity) {
            throw IllegalArgumentException("Only the current lender can initiate an IOU transfer")
        }

        val outputIOU = inputIOU.withNewLender(newLender)
        val command = Command(IOUContract.Commands.Transfer(), participants.map { it.owningKey })
        val outputStateAndContract = StateAndContract(outputIOU, IOUContract.IOU_CONTRACT_ID)
        val transactionBuilder = TransactionBuilder(notary = notary).withItems(
            inputStateAndRef, outputStateAndContract, command
        )

        return Pair(transactionBuilder, participants)
    }
}

/**
 * This is the flow which signs IOU transfers.
 * The signing is handled by the [SignTransactionFlow].
 */
@InitiatedBy(IOUTransferFlow::class)
class IOUTransferFlowResponder(val flowSession: FlowSession) : FlowLogic<Unit>() {
    @Suspendable
    override fun call() {
        val signedTransactionFlow = object : SignTransactionFlow(flowSession) {
            override fun checkTransaction(stx: SignedTransaction) = requireThat {
                val output = stx.tx.outputs.single().data
                "This must be an IOU transaction" using (output is IOUState)
            }
        }

        val signedTxID = subFlow(signedTransactionFlow).id
        subFlow(ReceiveFinalityFlow(flowSession, expectedTxId = signedTxID))
    }
}